### PR TITLE
Product for location

### PR DESCRIPTION
**Pull request by twymer**
Improve UI for @esoergel's products stocked by location feature. Still doesn't really do much of anything, and still behind a feature flag ..

Paired w/ @esoergel

### CODE REVIEW

**Comment by dimagimon** [comment]
:+1: No lint errors found.

File: corehq/apps/hqwebapp/models.py [hunk 1]
```diff
@@ -488,10 +487,6 @@ def sidebar_items(self):
                     'title': LocationFieldsView.page_name(),
                     'urlname': LocationFieldsView.urlname,
                 },
-                {
-                    'title': ProductsPerLocationView.page_title,
-                    'urlname': ProductsPerLocationView.urlname,
-                },
```
Comment (esoergel) [inline corehq/apps/hqwebapp/models.py]: We moved this functionality to the `EditLocationView`

File: corehq/apps/locations/models.py [hunk 2]
```diff
@@ -27,10 +28,36 @@ class SQLLocation(MPTTModel):
...
+            return SQLProduct.by_domain(self.domain)
+        else:
+            return self._products.all()
+
+    @products.setter
+    def products(self, value):
+        # this will set stocks_all_products to true if the user
+        # has added all products in the domain to this location
+        self.stocks_all_products = (set(value) ==
+                                    set(SQLProduct.by_domain(self.domain)))
```
Comment (esoergel) [inline corehq/apps/locations/models.py]: Basically, if the user selects all products for their location, any new products on that domain will also show up.  We debated this a bit, and this seems like the right combination of ease-of-implementation and intuitiveness for the user.

File: corehq/apps/locations/models.py [hunk 3]
```diff
@@ -388,6 +407,19 @@ def linked_supply_point(self):
...
+        """
+        Brute force lookup for the LocationType object
+        that corresponds to this locations type.
+        This could definitely use a more efficient way,
+        but no domains at this point have a large list of
+        types.
+        """
+        for loc_type in Domain.get_by_name(self.domain).location_types:
+            if loc_type.name == self.location_type:
+                return loc_type
```
Comment (esoergel) [inline corehq/apps/locations/models.py]: This is copypasta'd from @twymer's other branch which is mired in QA.  Keeping this abstraction so we can easily swap it out once a better way to do this exists.

File: corehq/apps/locations/templates/locations/manage/location.html [hunk 4]
```diff
@@ -4,14 +4,27 @@
 {% load hqstyle_tags %}
 {% load crispy_forms_tags %}
 
+{% block head %} {{ block.super }}
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/location.html]: most of the changes in this file have just been moved from the now defunct `locations/manage/products_per_location.html`.

File: corehq/apps/locations/templates/locations/manage/location.html [hunk 5]
```diff
@@ -60,30 +73,62 @@
...
-            <span class="uneditable-input">
-              {{ amount }}
-            </span>
-          </div>
-        </div>
-      {% endfor %}
+  <ul class="nav nav-tabs">
+    <li class="active"><a href="#basic-info" data-toggle="tab">{% trans "Basic" %}</a></li>
+
+    {% if products_per_location_form %}
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/location.html]: moved the toggle check to the view

File: corehq/apps/locations/templates/locations/manage/location.html [hunk 6]
```diff
@@ -60,30 +73,62 @@
...
-        <div class="control-group">
-          <label class="control-label">{{ code }}</label>
-          <div class="controls">
-            <span class="uneditable-input">
-              {{ amount }}
-            </span>
-          </div>
-        </div>
-      {% endfor %}
+  <ul class="nav nav-tabs">
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/location.html]: Added tabs to the location edit page, currently just "basic" and "products", but I think there's another in the works.

File: corehq/apps/locations/templates/locations/manage/locations.html [hunk 7]
```diff
@@ -139,12 +134,6 @@
                 {% endif %}
             </span>
 
-            {% if request|toggle_enabled:"PRODUCTS_PER_LOCATION" %}
-              <a class="loc_edit btn btn-primary" data-bind="attr: { href: loc_products_url(uuid()) }">
-                {% trans "Products" %}
-              </a>
-            {% endif %}
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/locations.html]: removed this link, as the edit UI is just on the normal location edit page now

File: corehq/apps/locations/tests/test_locations.py [hunk 8]
```diff
@@ -1,12 +1,73 @@
...
-from corehq.apps.commtrack.helpers import make_supply_point
+from corehq.apps.commtrack.helpers import make_supply_point, make_product
 from corehq.apps.users.models import CommCareUser
 from django.test import TestCase
+from corehq.apps.products.models import SQLProduct
 
 from corehq.apps.domain.shortcuts import create_domain
 
 
+class LocationProducts(TestCase):
```
Comment (esoergel) [inline corehq/apps/locations/tests/test_locations.py]: this tests the getter and setter

File: corehq/apps/locations/views.py [hunk 9]
```diff
@@ -166,6 +166,27 @@ def location_form(self):
         return LocationForm(self.location, is_new=True)
 
     @property
+    def all_products(self):
```
Comment (esoergel) [inline corehq/apps/locations/views.py]: a lot of this is moved from `ProductsPerLocationView`, which we deleted

File: corehq/apps/locations/views.py [hunk 10]
```diff
@@ -220,7 +262,10 @@ def unarchive_location(request, domain, loc_id):
     })
 
 
-class IndividualLocationMixin(object):
```
Comment (esoergel) [inline corehq/apps/locations/views.py]: merged this back in to `EditLocationView`, it's only remaining use

File: corehq/apps/locations/views.py [hunk 11]
```diff
@@ -179,20 +200,41 @@ def page_context(self):
...
 
+    def products_form_post(self, request, *args, **kwargs):
+        products = SQLProduct.objects.filter(
+            product_id__in=request.POST.getlist('selected_ids', [])
+        )
+        self.sql_location.products = products
+        self.sql_location.save()
+        return self.form_valid()
+
+    def post(self, request, *args, **kwargs):
```
Comment (esoergel) [inline corehq/apps/locations/views.py]: dispatch to separate post handlers each of the forms (one per tab).  This pattern is copied from `EditCommCareUserView`.

File: corehq/apps/users/forms.py [hunk 12]
```diff
@@ -367,7 +367,8 @@ class MultipleSelectionForm(forms.Form):
     def __init__(self, *args, **kwargs):
         self.helper = FormHelper()
         self.helper.form_tag = False
-        self.helper.add_input(Submit('submit', 'Update'))
+        submit_label = kwargs.pop('submit_label', "Update")
+        self.helper.add_input(Submit('submit', submit_label))
```
Comment (esoergel) [inline corehq/apps/users/forms.py]: let you customize the submit button for the `MultipleSelectionForm` (the one used for specifying which users are in a group)

**Comment by esoergel** [comment]
@twymer ruh roh: https://travis-ci.org/dimagi/commcare-hq/jobs/49364815

**Comment by twymer** [comment]
@esoergel Yeah I'm on it

**Comment by dimagimon** [comment]
:+1: No lint errors found.

File: corehq/apps/locations/views.py (hunk 11 above)
Comment (millerdev) [inline corehq/apps/locations/views.py]: Is it possible for the user to make changes on one tab, then switch to the other and make more changes before finally saving? Will all changes be saved in that case?

**Comment by millerdev** [comment]
Did a quick review and nothing big jumped out at me other than the one thing I commented on. I don't feel like I have enough context to merge this, sorry.

File: corehq/apps/locations/views.py (hunk 11 above)
Comment (twymer) [inline corehq/apps/locations/views.py]: @millerdev No, the tabs are separate (which is also true for tabs on other things like users, I think).

Comment (millerdev) [inline corehq/apps/locations/views.py]: :-(

**Comment by czue** [comment]
(i think if you guys paired feel free to "self merge")

**Pull request merged**